url72c32f3c54_Quickhull3d_quickhull3d_tgz-pJ8-com_github_quickhull3d_SimpleExampleJ8/src/com/github/quickhull3d/QuickHull3D.java

```diff
@@ -0,0 +1,1334 @@
+        boolean ccw = (flags & CLOCKWISE) == 0;
+        boolean indexedFromOne = (flags & INDEXED_FROM_ONE) != 0;
+        boolean pointRelative = (flags & POINT_RELATIVE) != 0;
+
+        HalfEdge hedge = face.he0;
+        int k = 0;
+        do {
+            int idx = hedge.head().index;
+            if (pointRelative) {
+                idx = vertexPointIndices[idx];
+            }
+            if (indexedFromOne) {
+                idx++;
+            }
+            indices[k++] = idx;
+            hedge = (ccw ? hedge.next : hedge.prev);
+        } while (hedge != face.he0);
+    }
+
+    protected void resolveUnclaimedPoints(FaceList newFaces) {
+        Vertex vtxNext = unclaimed.first();
+        for (Vertex vtx = vtxNext; vtx != null; vtx = vtxNext) {
+            vtxNext = vtx.next;
+
+            double maxDist = tolerance;
+            Face maxFace = null;
+            for (Face newFace = newFaces.first(); newFace != null; newFace = newFace.next) {
+                if (newFace.mark == Face.VISIBLE) {
+                    double dist = newFace.distanceToPlane(vtx.pnt);
+                    if (dist > maxDist) {
+                        maxDist = dist;
+                        maxFace = newFace;
+                    }
+                    if (maxDist > 1000 * tolerance) {
+                        break;
+                    }
+                }
+            }
+            if (maxFace != null) {
+                addPointToFace(vtx, maxFace);
+                if (LOG.isDebugEnabled() && vtx.index == findIndex) {
+                    LOG.debug(findIndex + " CLAIMED BY " + maxFace.getVertexString());
+                }
+            } else {
+                if (LOG.isDebugEnabled() && vtx.index == findIndex) {
+                    LOG.debug(findIndex + " DISCARDED");
+                }
+            }
+        }
+    }
+
+    protected void deleteFacePoints(Face face, Face absorbingFace) {
+        Vertex faceVtxs = removeAllPointsFromFace(face);
+        if (faceVtxs != null) {
+            if (absorbingFace == null) {
+                unclaimed.addAll(faceVtxs);
+            } else {
+                Vertex vtxNext = faceVtxs;
+                for (Vertex vtx = vtxNext; vtx != null; vtx = vtxNext) {
+                    vtxNext = vtx.next;
+                    double dist = absorbingFace.distanceToPlane(vtx.pnt);
+                    if (dist > tolerance) {
+                        addPointToFace(vtx, absorbingFace);
+                    } else {
+                        unclaimed.add(vtx);
+                    }
+                }
+            }
+        }
+    }
+
+    private static final int NONCONVEX_WRT_LARGER_FACE = 1;
+
+    private static final int NONCONVEX = 2;
+
+    protected double oppFaceDistance(HalfEdge he) {
+        return he.face.distanceToPlane(he.opposite.face.getCentroid());
+    }
+
+    private boolean doAdjacentMerge(Face face, int mergeType) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 15. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.